lib/doh.c

```diff
@@ -888,8 +996,238 @@ UNITTEST void de_cleanup(struct dohentry *d)
   for(i = 0; i < d->numcname; i++) {
     Curl_dyn_free(&d->cname[i]);
   }
+#ifdef USE_HTTPSRR
+  for(i = 0; i < d->numhttps_rrs; i++)
+    free(d->https_rrs[i].val);
+#endif
+}
+
+#ifdef USE_HTTPSRR
+
+/*
+ * @brief decode the DNS name in a binary RRData
+ * @param buf points to the buffer (in/out)
+ * @param remaining points to the remaining buffer length (in/out)
+ * @param dnsname returns the string form name on success
+ * @return is 1 for success, error otherwise
+ *
+ * The encoding here is defined in
+ * https://tools.ietf.org/html/rfc1035#section-3.1
+ *
+ * The input buffer pointer will be modified so it points to
+ * just after the end of the DNS name encoding on output. (And
+ * that's why it's an "unsigned char **" :-)
+ */
+static int local_decode_rdata_name(unsigned char **buf, size_t *remaining,
+                                   char **dnsname)
+{
+    unsigned char *cp = NULL;
+    size_t rem = 0;
+    char *thename = NULL, *tp = NULL;
+    unsigned char clen = 0; /* chunk len */
+
+    if(!buf || !remaining || !dnsname)
+      return CURLE_OUT_OF_MEMORY;
+    rem = *remaining;
+    thename = calloc(1, CURL_MAXLEN_host_name);
+    if(!thename)
+      return CURLE_OUT_OF_MEMORY;
+    cp = *buf;
+    tp = thename;
+    clen = *cp++;
+    if(clen == 0) {
+      /* special case - return "." as name */
+      thename[0] = '.';
+      thename[1] = 0x00;
+    }
+    while(clen) {
+      if(clen > rem) {
+        free(thename);
+          return CURLE_OUT_OF_MEMORY;
+      }
+      if(((tp - thename) + clen) > CURL_MAXLEN_host_name) {
+        free(thename);
+        return CURLE_OUT_OF_MEMORY;
+      }
+      memcpy(tp, cp, clen);
+      tp += clen;
+      *tp++ = '.';
+      cp += clen;
+      rem -= (clen + 1);
+      clen = *cp++;
+    }
+    *buf = cp;
+    *remaining = rem - 1;
+    *dnsname = thename;
+    return CURLE_OK;
+}
+
+static int local_decode_rdata_alpn(unsigned char *rrval, size_t len,
+                                   char **alpns)
+{
+  /*
+   * spec here is as per draft-ietf-dnsop-svcb-https, section-7.1.1
+   * encoding is catenated list of strings each preceded by a one
+   * octet length
+   * output is comma-sep list of the strings
+   * implementations may or may not handle quoting of comma within
+   * string values, so we migth see a comma within the wire format
+   * version of a string, in which case we'll precede that by a
+   * backslash - same goes for a backslash character
+   */
+  int remaining = (int) len;
```

Comment:
Would this not avoid the need for several typecasts if you make it a `size_t`  here?

Reply:
I think ``int remaining`` is maybe safer here - with some badly encoded ALPN values, ``remaining``  might go negative so not sure the typecasts are avoidable though doubtless the code could look nicer.